README.md
# Iron Broom

Cleanup docker containers, volumes and images in a [Rancher](http://rancher.com/) infrastructure.

This image will periodically clean up exited standalone containers and remove images and volumes that aren't in use by any container (a.k.a 'dangling'). If a container is part of a stack it will be left alone, as stack-managed containers should be started, stopped & removed by the stack.

Normally any Docker containers that exit are still kept on disk until `docker rm -v` is used to clean them up. Similarly any images that aren't used any more are kept around. For a cluster node that see lots of containers start and stop, large amounts of exited containers and old image versions can fill up the disk. A Gitlab CI multi runner starts a lot of containers and leaves them around.

# Flow

* find all terminated containers on the host which are dead longer than `DECAY_TIME`
* select containers without `io.rancher.stack.name` label
* delete them
* delete dangling volumes & images

# Environment Variables

The default parameters can be overridden by setting environment variables on the container using the docker run -e flag.

`CLEAN_PERIOD` (integer, default `1800`) sleep seconds between cleanup runs

`DECAY_TIME` (integer, default `30`) minutes a container must have been dead before it gets purged

`DEBUG` (boolean, default `false`) debug output

`PRETEND` (boolean, default `true`) do not actually delete, just do a dry-run and log

`LOOP` (boolean, default `true`) for testing purposes - execute just once and exit

# Deployment

This image is deployed best via [Rancher catalog](https://github.com/fjakop/rancher-catalog). Appropriate `docker-compose.yml` and `rancher-compose.yml` are included in catalog subfolder.

The image uses the Docker client to to list and remove containers and images. For this reason the Docker client and socket is mapped into the container.



This tool was inspired by https://github.com/meltwater/docker-cleanup


CHANGELOG.md
* DECAY_TIME specifies time in minutes for wich a container must have been dead before it gets pruned
* Use docker-ce cli-api (docker container|volume|image prune) with filtering facilities

**1.0** 2017-08-16

* Clean all terminated standalone containers, dangling volumes and dangling images


docker/run.sh
#!/bin/bash


if [ ! -e "/var/run/docker.sock" ]; then
  echo "=> Cannot find docker socket(/var/run/docker.sock), please check the command!"
  exit 1
fi

if docker version >/dev/null; then
  echo "docker is running properly, starting cleanup"
else
  echo "Cannot run docker binary at /usr/bin/docker"
  echo "Please check if the docker binary is mounted correctly"
  exit 1
fi

if ! [[ "$CLEAN_PERIOD" =~ ^[0-9]+$ ]]; then
  echo "=> CLEAN_PERIOD not defined, use the default value."
  CLEAN_PERIOD=1800
fi

if [ "${LOOP}" != "false" ]; then
  LOOP=true
fi

if [ "${PRETEND}" != "false" ]; then
  PRETEND=true
fi

if [ "${DEBUG}" == "0" ]; then
  unset DEBUG
fi

trap '{ echo "User Interupt."; exit 1; }' SIGINT
trap '{ echo "SIGTERM received, exiting."; exit 0; }' SIGTERM

while [ 1 ]
do

  # get all terminated containers
  TERMINATED=$(docker ps -a -q -f status=exited -f status=dead)
  if [ $DEBUG ]; then echo -e "Terminated container IDs\n$TERMINATED"; fi

  # handle all terminated containers
  for CONTAINER_ID in $TERMINATED; do
    CONTAINER_NAME=$(docker inspect --format='{{(index .Name)}}' $CONTAINER_ID | sed -e 's/^\s*\/\(.*\)$/\1/g')
    LABELS=$(docker inspect --format='{{.Config.Labels}}' $CONTAINER_ID)
    # check if container belongs to a stack
    if [[ ${LABELS} == *"io.rancher.stack.name:"* ]]; then
      if [ $DEBUG ]; then echo "Keeping terminated container $CONTAINER_NAME since it is contained in a stack"; fi
    else
      if [ $DEBUG ]; then echo "Removing terminated standalone container $CONTAINER_NAME"; fi
      [ "${PRETEND}" == "true" ] || docker rm $CONTAINER_NAME
    fi
  done
  unset CONTAINER_ID
  
  # remove dangling volumes
  VOLUMES=$(docker volume ls -qf dangling=true)
  if [ $DEBUG ]; then 
    echo "Removing dangling volumes"
    echo $VOLUMES
  fi
  [ "${PRETEND}" == "true" ] || echo $VOLUMES | xargs -r docker volume rm

  # remove dangling images
  IMAGES=$(docker images -qf dangling=true)
  if [ $DEBUG ]; then 
    echo "Removing dangling images"
    echo $IMAGES
  fi
  [ "${PRETEND}" == "true" ] || echo $IMAGES | xargs -r docker rmi

  # Run forever or exit after the first run depending on the value of $LOOP
  [ "${LOOP}" == "true" ] || break

  echo "=> Next clean will be started in ${CLEAN_PERIOD} seconds"
  sleep ${CLEAN_PERIOD} & wait
done


entrypoint.sh
#!/bin/bash


if [ ! -e "/var/run/docker.sock" ]; then
  echo "=> Cannot find docker socket(/var/run/docker.sock), please check the command!"
  exit 1
fi

if docker version >/dev/null; then
  echo "docker is running properly, starting cleanup"
else
  echo "Cannot run docker binary at /usr/bin/docker"
  echo "Please check if the docker binary is mounted correctly"
  exit 1
fi

if ! [[ "$CLEAN_PERIOD" =~ ^[0-9]+$ ]]; then
  echo "=> CLEAN_PERIOD not defined, use the default value."
  CLEAN_PERIOD=1800
fi

if ! [[ "$DECAY_TIME" =~ ^[0-9]+$ ]]; then
  echo "=> DECAY_TIME not defined, use the default value."
  DECAY_TIME=30
fi

if [ "${LOOP}" != "false" ]; then
  LOOP=true
fi

if [ "${PRETEND}" != "false" ]; then
  PRETEND=true
fi

if [ "${DEBUG}" == "false" ]; then
  unset DEBUG
fi

trap '{ echo "User Interupt."; exit 1; }' SIGINT
trap '{ echo "SIGTERM received, exiting."; exit 0; }' SIGTERM

while [ 1 ]
do

  # prune terminated containers without rancher stack exited longer than DECAY_TIME
  [ -n "$DEBUG" ] && echo "pruning terminated containers without rancher stack label exited longer than $DECAY_TIME minutes"
  [ "${PRETEND}" == "true" ] || docker container prune -f --filter "label!=io.rancher.stack.name" --filter "until=${DECAY_TIME}m"
  
  # prune unused volumes
  [ -n "$DEBUG" ] && echo "pruning dangling volumes"
  [ "${PRETEND}" == "true" ] || docker volume prune -f

  # remove dangling images
  [ -n "$DEBUG" ] && echo "pruning dangling images"
  [ "${PRETEND}" == "true" ] || docker image prune -f

  # Run forever or exit after the first run depending on the value of $LOOP
  [ "${LOOP}" == "true" ] || break

  echo "=> Next clean will be started in ${CLEAN_PERIOD} seconds"
  sleep ${CLEAN_PERIOD} & wait
done
